lib/Graph/Graph.cpp

```diff
@@ -1066,6 +1066,35 @@ LogNode *Function::createLog(llvm::StringRef name, NodeValue input,
   return addNode(new LogNode(name, outTy ? outTy : input.getType(), input));
 }
 
+Node *Function::createLogit(llvm::StringRef name, NodeValue data, float eps) {
+  // Create Splat node for the eps parameter.
+  auto *SE = createSplat(name.str() + ".splat", data.getType(), eps);
+
+  // Compute clamped x = max(min(x, 1 - eps), eps).
+  // First, compute 1 - eps.
+  auto *ones =
+      createSplat(name.str() + ".ones", SE->getResult().getType(), 1.0f);
+  auto *minArg = createSub(name.str() + ".sub", ones, SE);
+  // Take the min of data and 1 - eps.
+  auto *MinN = createMin(name.str() + ".min", data, minArg);
+  // Take the max of the min and eps (thus obtaining the clamped x).
+  auto *MaxN = createMax(name.str() + ".max", MinN, SE);
```

Comment:
There is an existing function `Function::createClip` that will clamp the input to a provided minimum and maximum. Consider using that instead, like so:

`auto CN = createClip(name.str() + ".clip", input, eps, 1-eps);`

`CN` will then replace `MaxN` as the input to the `SubNode` below.

Reply:
Nice! Clip operator would be much cleaner!

One aspect I'm wondering about is numerical reproducibility compared to Caffe2 (which also has Clip, but does not use it in its Logit implementation).
The difference is that Clip's implementation (in Caffe2 and in Glow) is a min-of-max (as in `std::min(std::max(v, lo), hi)`), whereas Logit's internal clamping implementation (in Caffe2) uses a max-of-min instead (as in `std::max(std::min(v, hi), lo)`).
In general, these are not commutative (https://2pi.dk/2016/05/ieee-min-max) although perhaps the inputs (like https://wandbox.org/permlink/iWTBzwmn4Y7y8Aka) are not reasonable enough to warrant special treatment (which a dedicated re-implementation may amount to). I'm curious, are there any reasonable values that would warrant it? If not I think Clip seems indeed preferable.

I'm wondering what's the best course of action here. Thoughts?

Reply:
Oh, I didn't realize they weren't commutative. Thanks for the link to that blog post!

It seems to me that the two clamping implementations are commutative only if `lo <= hi` is true. I would argue that is not  a problem because if `lo > hi`, the caller is using the function incorrectly.

Since the domain of the logit function is [0, 1], I l think it makes sense to restrict values for `eps` to [0, 0.5] because this is the largest subset of [0, 1] for which `eps <= 1-eps` is true. Maybe it's worth adding an `assert` to make sure `eps` is in this range.

Reply:
Hah, is it a coincidence that that blog post is written by @stoklund? :-D

Reply:
Wow, that Alive-NJ stuff is really cool.  I've got to play with that sometime!